lib/path.nix

```diff
@@ -0,0 +1,172 @@
+      #   part respectively
+      #
+      # These are the only cases where "." and "" parts can occur
+      parts = split "/+(\\./+)*" path;
+
+      # `split` creates a list of 2 * k + 1 elements, containing the k +
+      # 1 parts, interleaved with k matches where k is the number of
+      # (non-overlapping) matches. This calculation here gets the number of parts
+      # back from the list length
+      # floor( (2 * k + 1) / 2 ) + 1 == floor( k + 1/2 ) + 1 == k + 1
+      partCount = length parts / 2 + 1;
+
+      # To assemble the final list of components we want to:
+      # - Skip a potential leading ".", normalising "./foo" to "foo"
+      # - Skip a potential trailing "." or "", normalising "foo/" and "foo/." to
+      #   "foo"
+      skipStart = if head parts == "." then 1 else 0;
+      skipEnd = if last parts == "." || last parts == "" then 1 else 0;
+
+      # We can now know the length of the result by removing the number of
+      # skipped parts from the total number
+      componentCount = partCount - skipEnd - skipStart;
+
+    in
+      # Special case of a single "." path component. Such a case leaves a
+      # componentCount of -1 due to the skipStart/skipEnd not verifying that
+      # they don't refer to the same character
+      if path == "." then []
+
+      # And we can use this to generate the result list directly. Doing it this
+      # way over a combination of `filter`, `init` and `tail` makes it more
+      # efficient, because we don't allocate any intermediate lists
+      else genList (index:
+        # To get to the element we need to add the number of parts we skip and
+        # multiply by two due to the interleaved layout of `parts`
+        elemAt parts ((skipStart + index) * 2)
+      ) componentCount;
+
+  # joins subpath components together
+  joinSubpath = components:
+    # Always return relative paths with `./` as a prefix (./path.md#leading-dots-for-subpaths)
+    "./" +
+    # An empty string is not a valid subpath, so we need to return a `.` when we have no components
+    (if components == [] then "." else concatStringsSep "/" components);
+
+in /* No rec! Add dependencies on this file at the top. */ {
+
+  /* Normalise a subpath.
+
+  - Limit repeating `/` to a single one
+
+  - Remove redundant `.` components
+
+  - Error on empty strings
+
+  - Remove trailing `/` and `/.`
+
+  - Error on `..` path components
+
+  - Add leading `./`
+
+  Type:
+    subpath.normalise :: String -> String
+
+  Laws:
+
+  - (Idempotency) Normalising multiple times gives the same result:
+
+        subpath.normalise (subpath.normalise p) == subpath.normalise p
+
+  - (Uniqueness) There's only a single normalisation for a path:
+
+        subpath.normalise p != subpath.normalise q => $(realpath -ms ${p}) != $(realpath -ms ${q})
+
+  - Doesn't change the path according to `realpath -ms`:
+
+        $(realpath -ms ${p}) == $(realpath -ms ${subpath.normalise p})
```

Comment:
I'm considering this future work, since there's no easy way to check this property (with `realpath`) with the current infrastructure.

I'm not sure if it's worth doing property checks, the 12 unit tests are fairly extensive. I'll add it if you think it's worth it though.

Reply:
> the current infrastructure.

You could duplicate the `lib/tests/sources.sh` approach, which is called from `release.nix` in the same directory.

If you want to test with a large amount of paths, you could batch them up, but only on the Nix side of the equation. `realpath` is a small C program so that shouldn't be too costly.

I would appreciate it, and it adds more credence to the stated laws.

Reply:
How about https://github.com/NixOS/nixpkgs/pull/205190/commits/284a71b3ed78135bb92c4fa38d20ba41dd432c05